File: 1/2020.11.01/718.Maximum Length of Repeated Subarray/java/solution.java
class Solution {
    public int findLength(int[] A, int[] B) {
        // dp
        int m = A.length, n = B.length;
        int[][] dp = new int[m+1][n+1];
        int res = 0;
        for (int i = m-1; i >= 0; i--) {
            for (int j = n-1; j >=0; j--) {
                if (A[i] == B[j]) {
                    dp[i][j] = dp[i+1][j+1] + 1;
                    res = Math.max(res, dp[i][j]);
                } else {
                    dp[i][j] = 0;
                }
            }
        }
        return res;
    }
}

File: 1/2020.11.01/718.Maximum Length of Repeated Subarray/python/solution.py
class Solution:
    def findLength(self, A: List[int], B: List[int]) -> int:
        def max_length(add_a, add_b, length):
            res = k = 0
            for i in range(length):
                if A[add_a + i] == B[add_b + i]:
                    k += 1
                    res = max(res, k)
                else:
                    k = 0
            return res
        m, n = len(A), len(B)
        res = 0
        for i in range(m):
            length = min(n, m-i)
            res = max(res, max_length(i, 0, length))
        for i in range(n):
            length = min(m, n-i)
            res = max(res, max_length(0, i, length))
        return res
